### PR TITLE
hedgedoc: 1.9.7 -> 1.9.8 and redo package

### PR DESCRIPTION
**Pull request by dasJ**
- Yarnify
- Do a full source build
- Clean up the wrapper things

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by dasJ** [comment]
This is a draft because I only roughly tested this. Feel free to undraft once tested more. Also, I pushed this directly into the repo rather than a fork so everyone can contribute ;)

**Comment by dasJ** [comment]
@ofborg test hedgedoc

File: pkgs/servers/web-apps/hedgedoc/default.nix [hunk 1]
```diff
@@ -1,73 +1,95 @@
...
+  ];
+  extraBuildInputs = [ python3 ];
 
-  configurePhase = ''
-    cp -r "$node_modules" node_modules
-    chmod -R u+w node_modules
-  '';
+  dontConfigure = true;
 
+  HOME = "/build";
```
Comment (SuperSandro2000) [inline pkgs/servers/web-apps/hedgedoc/default.nix:58]:
```suggestion
  HOME = "$TMPDIR";
```

File: pkgs/servers/web-apps/hedgedoc/default.nix [hunk 2]
```diff
@@ -1,73 +1,95 @@
...
-  offlineCache = fetchYarnDeps {
-    yarnLock = src + "/yarn.lock";
-    sha256 = "0qkc26ks33vy00jgpv4445wzgxx1mzi70pkm1l8y9amgd9wf9aig";
-  };
+  nativeBuildInputs = [
+    makeBinaryWrapper
+    yarn
+    python3 # needed for sqlite node-gyp
+  ];
+  extraBuildInputs = [ python3 ];
```
Comment (SuperSandro2000) [inline pkgs/servers/web-apps/hedgedoc/default.nix:54]: Has this any effect for mkDerivation?

File: pkgs/servers/web-apps/hedgedoc/default.nix [hunk 3]
```diff
@@ -1,73 +1,95 @@
...
+    inherit src;
+
+    nativeBuildInputs = [
+      cacert # needed for git
+      gitMinimal # needed for a dep
+      nodePackages.npm # needed for a dep
+      yarn
+    ];
+
+    HOME = "/build";
```
Comment (SuperSandro2000) [inline pkgs/servers/web-apps/hedgedoc/default.nix:34]:
```suggestion
    HOME = "$TMPDIR";
```

File: pkgs/servers/web-apps/hedgedoc/default.nix (hunk 2 above)
Comment (dasJ) [reply pkgs/servers/web-apps/hedgedoc/default.nix:54]: Nope, it's a leftover from the old package

File: pkgs/servers/web-apps/hedgedoc/default.nix [hunk 4]
```diff
@@ -1,73 +1,94 @@
...
-  nativeBuildInputs = [ which makeWrapper ];
-  extraBuildInputs = [ python3 ];
+  offlineCache = stdenv.mkDerivation {
+    name = "hedgedoc-${version}-offline-cache";
+    inherit src;
+
+    nativeBuildInputs = [
+      cacert # needed for git
+      gitMinimal # needed for a dep
+      nodePackages.npm # needed for a dep
```
Comment (SuperSandro2000) [inline pkgs/servers/web-apps/hedgedoc/default.nix:30]: This is also in nodejs directly but if it works, it works

**Review by SuperSandro2000 — APPROVED** [review]
Tested locally with ``CMD_DOMAIN=localhost CMD_DB_DIALECT=sqlite CMD_PORT=9000 CMD_URL_ADDPORT=true ./result/bin/hedgedoc`` and it works.

File: pkgs/servers/web-apps/hedgedoc/default.nix [hunk 5]
```diff
@@ -1,73 +1,94 @@
...
+      yarn
+    '';
+
+    outputHashMode = "recursive";
+    outputHash = "sha256-4TPc9J9u0rO9WfZVbFWwVAM5WY3n5zJdhIlNYiZNRoM=";
+  };
 
-  packageJSON = ./package.json;
-  yarnFlags = [ "--production" ];
+in stdenv.mkDerivation rec {
```
Comment (drupol) [inline pkgs/servers/web-apps/hedgedoc/default.nix:45]: How about using a better overriding practice like the `finalAttrs` pattern?

```suggestion
in stdenv.mkDerivation (finalAttrs: {
```

**Review by drupol — CHANGES_REQUESTED** [review]
Glad to see this update! I didn't know it was possible with the current state of toolset.

I added a minor comment, it would be nice to implement the change.

Thanks!

File: pkgs/servers/web-apps/hedgedoc/default.nix (hunk 5 above)
Comment (dasJ) [reply pkgs/servers/web-apps/hedgedoc/default.nix:45]: I don't know if I will have time in the next couple of days, if anyone is interested, they can just push the branch so this gets through faster

Comment (drupol) [reply pkgs/servers/web-apps/hedgedoc/default.nix:45]: Simply remove the rec and then it's fine.

**Review by SuperSandro2000 — APPROVED** [review]

**Review by drupol — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
Not sure why the cache is suddenly no longer reproducible

```
➤ 8;;https://yarnpkg.com/advanced/error-codes#yn0013---fetch_not_cachedYN00138;;: │ zip-stream@npm:4.1.0 can't be found in the cache and will be fetched from the remote registry
➤ 8;;https://yarnpkg.com/advanced/error-codes#yn0013---fetch_not_cachedYN00138;;: │ yargs@npm:16.2.0 can't be found in the cache and will be fetched from the remote registry
➤ 8;;https://yarnpkg.com/advanced/error-codes#yn0013---fetch_not_cachedYN00138;;: │ yeast@npm:0.1.2 can't be found in the cache and will be fetched from the remote registry
➤ 8;;https://yarnpkg.com/advanced/error-codes#yn0013---fetch_not_cachedYN00138;;: │ yocto-queue@npm:0.1.0 can't be found in the cache and will be fetched from the remote registry
➤ 8;;https://yarnpkg.com/advanced/error-codes#yn0013---fetch_not_cachedYN00138;;: │ zip-stream@npm:4.1.0 can't be found in the cache and will be fetched from the remote registry
➤ 8;;https://yarnpkg.com/advanced/error-codes#yn0013---fetch_not_cachedYN00138;;: │ zwitch@npm:2.0.4 can't be found in the cache and will be fetched from the remote registry
➤ 8;;https://yarnpkg.com/advanced/error-codes#yn0013---fetch_not_cachedYN00138;;: │ yeast@npm:0.1.2 can't be found in the cache and will be fetched from the remote registry
➤ 8;;https://yarnpkg.com/advanced/error-codes#yn0013---fetch_not_cachedYN00138;;: │ yocto-queue@npm:0.1.0 can't be found in the cache and will be fetched from the remote registry
➤ 8;;https://yarnpkg.com/advanced/error-codes#yn0013---fetch_not_cachedYN00138;;: │ zip-stream@npm:4.1.0 can't be found in the cache and will be fetched from the remote registry
➤ 8;;https://yarnpkg.com/advanced/error-codes#yn0013---fetch_not_cachedYN00138;;: │ zwitch@npm:2.0.4 can't be found in the cache and will be fetched from the remote registry
➤ 8;;https://yarnpkg.com/advanced/error-codes#yn0013---fetch_not_cachedYN00138;;: │ meta-marked@https://github.com/hedgedoc/meta-marked.git#commit=0cb5065253ab0f9851baec34dc9edbb40eabf3d0 can't be found in the cache and will be fetched from the remote repository
➤ 8;;https://yarnpkg.com/advanced/error-codes#yn0018---cache_checksum_mismatchYN00188;;: │ meta-marked@https://github.com/hedgedoc/meta-marked.git#commit=0cb5065253ab0f9851baec34dc9edbb40eabf3d0: The remote archive doesn't match the expected checksum
➤ YN0000: └ Completed in 11s 602ms
➤ YN0000: Failed with errors in 11s 821ms
error: builder for '/nix/store/l9awjs5iprv9h4yya9dwvhqjsmcx1i5z-hedgedoc-1.9.8-offline-cache.drv' failed with exit code 1;
```

File: pkgs/servers/web-apps/hedgedoc/default.nix [hunk 6]
```diff
@@ -1,73 +1,95 @@
...
+    inherit src;
+
+    nativeBuildInputs = [
+      cacert # needed for git
+      gitMinimal # needed for a dep
+      nodePackages.npm # needed for a dep
+      yarn
+    ];
+
+    HOME = "$TMPDIR";
```
Comment (SuperSandro2000) [inline pkgs/servers/web-apps/hedgedoc/default.nix:35]: Changing that broke the FOD, going to change it back. I also changed npm to use nodejs, to not rely on nodePackages while I was testing around.

**Review by SuperSandro2000 — APPROVED** [review]
Finally could deploy to a live instance. Thank you so much.

**Review by drupol — APPROVED** [review]

**Comment by drupol** [comment]
The `offline-cache` for `aarch64-linux` platform is different.

image

Do you think we can fix this?

File: pkgs/servers/web-apps/hedgedoc/default.nix [hunk 7]
```diff
@@ -1,73 +1,94 @@
...
+  src = fetchFromGitHub {
+    owner = "hedgedoc";
+    repo = "hedgedoc";
+    rev = version;
+    hash = "sha256-gp1TeYHwH7ffaSMifdURb2p+U8u6Xs4JU4b4qACEIDw=";
   };
 
-  nativeBuildInputs = [ which makeWrapper ];
-  extraBuildInputs = [ python3 ];
+  offlineCache = stdenv.mkDerivation {
```
Comment (yu-re-ka) [inline pkgs/servers/web-apps/hedgedoc/default.nix:23]: Please don't roll your own yarn cache generation, we have fetchYarnDeps for this which is safe against impure things and changes to yarn

Comment (drupol) [reply pkgs/servers/web-apps/hedgedoc/default.nix:23]: Ha ! That might fix the issue I mentioned earlier with different hashes produced ?

**Comment by yu-re-ka** [comment]
Ah, this is a yarn 2 / berry lockfile. we don't have a solution for those yet. please add a comment that this is why fetchYarnDeps is not used.

According to this documentation here it can be told to download dependencies for a list of supported platforms: https://yarnpkg.com/configuration/yarnrc#supportedArchitectures
Which should fix the hash mismatch

**Review by drupol — APPROVED** [review]

**Pull request merged**